diff --git a/test/mpi/rma/reqops.c b/test/mpi/rma/reqops.c
index ef2636f..f4509ca 100644
--- a/test/mpi/rma/reqops.c
+++ b/test/mpi/rma/reqops.c
@@ -86,6 +86,7 @@ int main( int argc, char *argv[] )
             errors++;
         }
 
+        MPI_Win_flush(0, window);
         if (rank < nproc-1) {
             MPI_Send(NULL, 0, MPI_BYTE, rank+1, 0, MPI_COMM_WORLD);
         }
@@ -125,6 +126,7 @@ int main( int argc, char *argv[] )
             errors++;
         }
 
+        MPI_Win_flush(0, window);
         if (rank < nproc-1) {
             MPI_Send(NULL, 0, MPI_BYTE, rank+1, 0, MPI_COMM_WORLD);
         }
@@ -164,6 +166,7 @@ int main( int argc, char *argv[] )
             errors++;
         }
 
+        MPI_Win_flush(0, window);
         if (rank < nproc-1) {
             MPI_Send(NULL, 0, MPI_BYTE, rank+1, 0, MPI_COMM_WORLD);
         }
