Index: src/pm/mpd/mpdlib.py
===================================================================
--- src/pm/mpd/mpdlib.py	(revision 3431)
+++ src/pm/mpd/mpdlib.py	(working copy)
@@ -4,6 +4,11 @@
 #       See COPYRIGHT in top-level directory.
 #
 
+# workaround to suppress deprecated module warnings in python2.6
+# see https://trac.mcs.anl.gov/projects/mpich2/ticket/362 for tracking
+import warnings
+warnings.filterwarnings('ignore', '.*the md5 module is deprecated.*', DeprecationWarning)
+warnings.filterwarnings('ignore', '.*the popen2 module is deprecated.*', DeprecationWarning)
 
 import sys, os, signal, popen2, socket, select, inspect
 
