diff --git a/confdb/aclocal_cc.m4 b/confdb/aclocal_cc.m4
index fd7152b..2ae1ea3 100644
--- a/confdb/aclocal_cc.m4
+++ b/confdb/aclocal_cc.m4
@@ -354,22 +354,49 @@ if test "$pac_cv_prog_c_weak_symbols" != "no" ; then
 fi
 AC_CACHE_CHECK([whether __attribute__ ((weak)) allowed],
 pac_cv_attr_weak,[
-AC_TRY_COMPILE([int foo(int) __attribute__ ((weak));],[int a;],
-pac_cv_attr_weak=yes,pac_cv_attr_weak=no)])
+AC_COMPILE_IFELSE([
+ AC_LANG_SOURCE([
+int foo(int) __attribute__ ((weak));
+
+int main(int argc, char *argv[]) {
+  int a;
+  return 0;
+}
+ ]),
+],pac_cv_attr_weak=yes,pac_cv_attr_weak=no)])
 # Note that being able to compile with weak_import doesn't mean that
 # it works.
 AC_CACHE_CHECK([whether __attribute__ ((weak_import)) allowed],
 pac_cv_attr_weak_import,[
-AC_TRY_COMPILE([int foo(int) __attribute__ ((weak_import));],[int a;],
-pac_cv_attr_weak_import=yes,pac_cv_attr_weak_import=no)])
+AC_COMPILE_IFELSE([
+ AC_LANG_SOURCE([
+int foo(int) __attribute__ ((weak_import));
+
+int main(int argc, char **argv) {
+  int a;
+  return 0;
+}
+ ]),
+],pac_cv_attr_weak_import=yes,pac_cv_attr_weak_import=no)])
 # Check if the alias option for weak attributes is allowed
 AC_CACHE_CHECK([whether __attribute__((weak,alias(...))) allowed],
 pac_cv_attr_weak_alias,[
 PAC_PUSH_FLAG([CFLAGS])
 # force an error exit if the weak attribute isn't understood
 CFLAGS=-Werror
-AC_TRY_COMPILE([int __foo(int a){return 0;} int foo(int) __attribute__((weak,alias("__foo")));],[int a;],
-pac_cv_attr_weak_alias=yes,pac_cv_attr_weak_alias=no)
+AC_COMPILE_IFELSE([
+ AC_LANG_SOURCE([
+int __foo(int a) {
+  return 0;
+}
+int foo(int) __attribute__((weak,alias("__foo")));
+
+int main(int argc, char **argv) {
+  int a;
+  return 0;
+}
+ ]),
+],pac_cv_attr_weak_alias=yes,pac_cv_attr_weak_alias=no)
 # Restore original CFLAGS
 PAC_POP_FLAG([CFLAGS])])
 if test "$pac_cv_attr_weak_alias" = "yes" ; then
@@ -396,7 +423,14 @@ AC_DEFUN([PAC_PROG_CC_WORKS],
 AC_MSG_CHECKING([whether the C compiler sets its return status correctly])
 AC_LANG_SAVE
 AC_LANG_C
-AC_TRY_COMPILE(,[int a = bzzzt;],notbroken=no,notbroken=yes)
+AC_COMPILE_IFELSE([
+ AC_LANG_SOURCE([
+int main (int argc, char **argv) {
+  int a = bzzzt;
+  return 0;
+}
+ ]),
+],notbroken=no,notbroken=yes)
 AC_MSG_RESULT($notbroken)
 if test "$notbroken" = "no" ; then
     AC_MSG_ERROR([installation or configuration problem: C compiler does not
@@ -521,7 +555,7 @@ if test "$enable_strict_done" != "yes" ; then
     # compiler.
     #   -Wno-type-limits -- There are places where we compare an unsigned to 
     #	    a constant that happens to be zero e.g., if x is unsigned and 
-    #	    MIN_VAL is zero, we'd like to do "MPIR_Assert(x >= MIN_VAL);".
+    #	    MIN_VAL is zero, we'd like to do "MPIU_Assert(x >= MIN_VAL);".
     #       Note this option is not supported by gcc 4.2.  This needs to be added 
     #	    after most other warning flags, so that we catch a gcc bug on 32-bit 
     #	    that doesn't give a warning that this is unsupported, unless another
@@ -671,7 +705,7 @@ dnl
 dnl D*/
 AC_DEFUN([PAC_ARG_STRICT],[
 AC_ARG_ENABLE(strict,
-	AC_HELP_STRING([--enable-strict], [Turn on strict compilation testing]))
+	AS_HELP_STRING([--enable-strict], [Turn on strict compilation testing]))
 PAC_CC_STRICT($enable_strict)
 CFLAGS="$CFLAGS $pac_cc_strict_flags"
 export CFLAGS
@@ -1249,9 +1283,16 @@ dnl D*/
 AC_DEFUN([PAC_FUNC_NEEDS_DECL],[
 AC_CACHE_CHECK([whether $2 needs a declaration],
 pac_cv_func_decl_$2,[
-AC_TRY_COMPILE([$1
-void (*fptr)(void) = (void(*)(void))$2;],[],
-pac_cv_func_decl_$2=no,pac_cv_func_decl_$2=yes)])
+AC_COMPILE_IFELSE([
+ AC_LANG_SOURCE([
+$1
+void (*fptr)(void) = (void(*)(void))$2;
+
+int main(int argc, char **argv) {
+  return 0;
+}
+ ]),
+]pac_cv_func_decl_$2=no,pac_cv_func_decl_$2=yes)])
 if test "$pac_cv_func_decl_$2" = "yes" ; then
 changequote(<<,>>)dnl
 define(<<PAC_FUNC_NAME>>, translit(NEEDS_$2_DECL, [a-z *], [A-Z__]))dnl
@@ -1279,12 +1320,28 @@ AC_REQUIRE([AC_PROG_CC_GNU])
 if test "$ac_cv_prog_gcc" = "yes" ; then
     AC_CACHE_CHECK([whether __attribute__ allowed],
 pac_cv_gnu_attr_pure,[
-AC_TRY_COMPILE([int foo(int) __attribute__ ((pure));],[int a;],
-pac_cv_gnu_attr_pure=yes,pac_cv_gnu_attr_pure=no)])
+AC_COMPILE_IFELSE([
+ AC_LANG_SOURCE([
+int foo(int) __attribute__ ((pure));
+
+int main(int argc, char **argv) {
+  int a;
+  return 0;
+}
+ ]),
+],pac_cv_gnu_attr_pure=yes,pac_cv_gnu_attr_pure=no)])
 AC_CACHE_CHECK([whether __attribute__((format)) allowed],
 pac_cv_gnu_attr_format,[
-AC_TRY_COMPILE([int foo(char *,...) __attribute__ ((format(printf,1,2)));],[int a;],
-pac_cv_gnu_attr_format=yes,pac_cv_gnu_attr_format=no)])
+AC_COMPILE_IFELSE([
+ AC_LANG_SOURCE([
+int foo(char *,...) __attribute__ ((format(printf,1,2)));
+
+int main(int argc, char **argv) {
+  int a;
+  return 0;
+}
+ ]),
+],pac_cv_gnu_attr_format=yes,pac_cv_gnu_attr_format=no)])
     if test "$pac_cv_gnu_attr_pure" = "yes" -a "$pac_cv_gnu_attr_format" = "yes" ; then
         AC_DEFINE(HAVE_GCC_ATTRIBUTE,1,[Define if GNU __attribute__ is supported])
     fi
@@ -1502,7 +1559,7 @@ AC_DEFUN([PAC_STRUCT_ALIGNMENT],[
 	is_largest=1
 
 	# See if long double exists
-	AC_TRY_COMPILE(,[long double a;],have_long_double=yes,have_long_double=no)
+	AC_CHECK_TYPE(long double,have_long_double=yes,have_long_double=no)
 
 	# Get sizes of regular types
 	AC_CHECK_SIZEOF(char)
diff --git a/confdb/aclocal_coverage.m4 b/confdb/aclocal_coverage.m4
index ce949d4..5edf664 100644
--- a/confdb/aclocal_coverage.m4
+++ b/confdb/aclocal_coverage.m4
@@ -18,7 +18,7 @@ AC_ARG_VAR([GCOV],[name/path for the gcov utility])
 AC_CHECK_PROGS([GCOV],[gcov])
 
 AC_ARG_ENABLE([coverage],
-              [AC_HELP_STRING([--enable-coverage],
+              [AS_HELP_STRING([--enable-coverage],
                               [Turn on coverage analysis using gcc and gcov])],
               [],[enable_coverage=no])
 
diff --git a/confdb/aclocal_cxx.m4 b/confdb/aclocal_cxx.m4
index 72febd8..f4139dc 100644
--- a/confdb/aclocal_cxx.m4
+++ b/confdb/aclocal_cxx.m4
@@ -25,11 +25,18 @@ AC_DEFUN([AX_CXX_BOOL],
 ac_cv_cxx_bool,
 [AC_LANG_SAVE
  AC_LANG_CPLUSPLUS
- AC_TRY_COMPILE([
+ AC_COMPILE_IFELSE([
+  AC_LANG_SOURCE([
 int f(int  x){return 1;}
 int f(char x){return 1;}
 int f(bool x){return 1;}
-],[bool b = true; return f(b);],
+
+int main(int argc, char **argv) {
+  bool b = true;
+  return f(b);
+}
+  ])
+ ],
  ac_cv_cxx_bool=yes, ac_cv_cxx_bool=no)
  AC_LANG_RESTORE
 ])
@@ -45,7 +52,17 @@ AC_DEFUN([AX_CXX_EXCEPTIONS],
 ac_cv_cxx_exceptions,
 [AC_LANG_SAVE
  AC_LANG_CPLUSPLUS
- AC_TRY_COMPILE(,[try { throw  1; } catch (int i) { return i; }],
+ AC_COMPILE_IFELSE([
+  AC_LANG_SOURCE([
+int main(int argc, char **argv) {
+  try {
+    throw  1;
+   } catch (int i) {
+     return i;
+   }
+}
+  ])
+ ],
  ac_cv_cxx_exceptions=yes, ac_cv_cxx_exceptions=no)
  AC_LANG_RESTORE
 ])
@@ -61,8 +78,20 @@ AC_DEFUN([AX_CXX_NAMESPACES],
 ac_cv_cxx_namespaces,
 [AC_LANG_SAVE
  AC_LANG_CPLUSPLUS
- AC_TRY_COMPILE([namespace Outer { namespace Inner { int i = 0; }}],
-                [using namespace Outer::Inner; return i;],
+ AC_COMPILE_IFELSE([
+  AC_LANG_SOURCE([
+namespace Outer {
+  namespace Inner {
+    int i = 0;
+  }
+}
+
+int main(int argc, char **argv) {
+  using namespace Outer::Inner;
+  return i;
+}
+  ])
+ ],
  ac_cv_cxx_namespaces=yes, ac_cv_cxx_namespaces=no)
  AC_LANG_RESTORE
 ])
@@ -81,10 +110,17 @@ ac_cv_cxx_namespace_std,
 if test "$ac_cv_cxx_namespaces" = yes ; then 
    AC_LANG_SAVE
    AC_LANG_CPLUSPLUS
-   AC_TRY_COMPILE([
+   AC_COMPILE_IFELSE([
+     AC_LANG_SOURCE([
 #include <iostream>
-using namespace std;],
-                [cout << "message\n";],
+using namespace std;
+
+int main(int argc, char **argv) {
+  cout << "message\n";
+  return 0;
+}
+  ])
+ ],
  ac_cv_cxx_namespace_std=yes, ac_cv_cxx_namespace_std=no)
    AC_LANG_RESTORE
 fi
diff --git a/confdb/aclocal_f77old.m4 b/confdb/aclocal_f77old.m4
index 5d6e2d9..fa1d6a9 100644
--- a/confdb/aclocal_f77old.m4
+++ b/confdb/aclocal_f77old.m4
@@ -77,7 +77,7 @@ $fxx_module
 EOF
     found_answer="no"
     if test -z "$ac_fcompilelink" ; then
-        ac_fcompilelink="${F77-f77} -o conftest $FFLAGS $flags conftest.f $LDFLAGS $LIBS 1>&AC_FD_CC"
+        ac_fcompilelink="${F77-f77} -o conftest $FFLAGS $flags conftest.f $LDFLAGS $LIBS 1>&AS_MESSAGE_LOG_FD"
     fi
     AC_MSG_CHECKING([whether ${F77-f77} $flags $libs works with GETARG and IARGC])
     if AC_TRY_EVAL(ac_fcompilelink) && test -x conftest ; then
@@ -124,7 +124,7 @@ EOF
         program main
         end
 EOF
-    ac_fcompilelink_test='${F77-f77} -o conftest $FFLAGS conftest.f $LDFLAGS $libs $LIBS 1>&AC_FD_CC'
+    ac_fcompilelink_test='${F77-f77} -o conftest $FFLAGS conftest.f $LDFLAGS $libs $LIBS 1>&AS_MESSAGE_LOG_FD'
     for libs in $save_trial_LIBS ; do
 	if test "$libs" = "0" ; then
 	    lib_ok="yes"
@@ -211,7 +211,7 @@ EOF
 	    AC_MSG_CHECKING([whether Fortran 77 routine names are case-insensitive $flagval])
 	    dnl we can use double quotes here because all is already
             dnl evaluated
-            ac_fcompilelink_test="${F77-f77} -o conftest $fflag $FFLAGS conftest.f $LDFLAGS $LIBS 1>&AC_FD_CC"
+            ac_fcompilelink_test="${F77-f77} -o conftest $fflag $FFLAGS conftest.f $LDFLAGS $LIBS 1>&AS_MESSAGE_LOG_FD"
 	    if AC_TRY_EVAL(ac_fcompilelink_test) && test -x conftest ; then
 	        AC_MSG_RESULT(yes)
 	    else
@@ -331,7 +331,7 @@ EOF
                 AC_MSG_CHECKING([whether ${F77-f77} $flags $libs works with $MSG])
 		IFS="$save_IFS"
 		dnl We need this here because we've fiddled with IFS
-	        ac_fcompilelink_test="${F77-f77} -o conftest $FFLAGS $flags conftest.f $LDFLAGS $libs $LIBS 1>&AC_FD_CC"
+	        ac_fcompilelink_test="${F77-f77} -o conftest $FFLAGS $flags conftest.f $LDFLAGS $libs $LIBS 1>&AS_MESSAGE_LOG_FD"
 		found_answer="no"
                 if AC_TRY_EVAL(ac_fcompilelink_test) && test -x conftest ; then
 		    if test "$ac_cv_prog_f77_cross" != "yes" -a \	 
@@ -353,8 +353,8 @@ EOF
 		    break
 	        else
                     AC_MSG_RESULT([no])
-		    echo "configure: failed program was:" >&AC_FD_CC
-                    cat conftest.f >&AC_FD_CC
+		    echo "configure: failed program was:" >&AS_MESSAGE_LOG_FD
+                    cat conftest.f >&AS_MESSAGE_LOG_FD
 	        fi
             done
         done
diff --git a/confdb/aclocal_libs.m4 b/confdb/aclocal_libs.m4
index 8868e99..5d9c941 100644
--- a/confdb/aclocal_libs.m4
+++ b/confdb/aclocal_libs.m4
@@ -8,7 +8,7 @@ dnl characters in it.  Use AS_TR_SH (and possibly AS_VAR_* macros) to handle
 dnl this case if it ever arises.
 AC_DEFUN([PAC_SET_HEADER_LIB_PATH],[
     AC_ARG_WITH([$1],
-                [AC_HELP_STRING([--with-$1=PATH],
+                [AS_HELP_STRING([--with-$1=PATH],
                                 [specify path where $1 include directory and lib directory can be found])],
 
                 [AS_CASE(["$withval"],
@@ -17,7 +17,7 @@ AC_DEFUN([PAC_SET_HEADER_LIB_PATH],[
                           with_$1=""])],
                 [with_$1=$2])
     AC_ARG_WITH([$1-include],
-                [AC_HELP_STRING([--with-$1-include=PATH],
+                [AS_HELP_STRING([--with-$1-include=PATH],
                                 [specify path where $1 include directory can be found])],
                 [AS_CASE(["$withval"],
                          [yes|no|''],
@@ -25,7 +25,7 @@ AC_DEFUN([PAC_SET_HEADER_LIB_PATH],[
                           with_$1_include=""])],
                 [])
     AC_ARG_WITH([$1-lib],
-                [AC_HELP_STRING([--with-$1-lib=PATH],
+                [AS_HELP_STRING([--with-$1-lib=PATH],
                                 [specify path where $1 lib directory can be found])],
                 [AS_CASE(["$withval"],
                          [yes|no|''],
@@ -47,11 +47,8 @@ AC_DEFUN([PAC_SET_HEADER_LIB_PATH],[
           [AS_IF([test -n "${with_$1}"],
                  dnl is adding lib64 by default really the right thing to do?  What if
                  dnl we are on a 32-bit host that happens to have both lib dirs available?
-                 [PAC_APPEND_FLAG([-L${with_$1}/lib],[LDFLAGS])
-                  AS_IF([test -d "${with_$1}/lib64"],
-		        [PAC_APPEND_FLAG([-L${with_$1}/lib64],[LDFLAGS])])
-                 ])
-          ])
+                 [PAC_APPEND_FLAG([-L${with_$1}/lib64],[LDFLAGS])
+                  PAC_APPEND_FLAG([-L${with_$1}/lib],[LDFLAGS])])])
 ])
 
 
diff --git a/confdb/aclocal_mpi.m4 b/confdb/aclocal_mpi.m4
index f0013a9..6bdfe88 100644
--- a/confdb/aclocal_mpi.m4
+++ b/confdb/aclocal_mpi.m4
@@ -498,9 +498,18 @@ AC_DEFUN([PAC_MPI_F2C],[
 AC_CACHE_CHECK([for MPI F2C and C2F routines],
 pac_cv_mpi_f2c,
 [
-AC_TRY_LINK([#include "mpi.h"],
-[MPI_Request request;MPI_Fint a;a = MPI_Request_c2f(request);],
-pac_cv_mpi_f2c="yes",pac_cv_mpi_f2c="no")
+AC_COMPILE_IFELSE([
+ AC_LANG_SOURCE([
+#include "mpi.h"
+
+int main (int argc, char **argv) {
+  MPI_Request request;
+  MPI_Fint a;
+  a = MPI_Request_c2f(request);
+  return 0;
+}
+ ]),
+], pac_cv_mpi_f2c="yes",pac_cv_mpi_f2c="no")
 ])
 if test "$pac_cv_mpi_f2c" = "yes" ; then 
     AC_DEFINE(HAVE_MPI_F2C,1,[Define if MPI has F2C]) 
diff --git a/confdb/aclocal_romio.m4 b/confdb/aclocal_romio.m4
index 64091ba..12052af 100644
--- a/confdb/aclocal_romio.m4
+++ b/confdb/aclocal_romio.m4
@@ -541,7 +541,7 @@ if test -z "$ac_f90ext" ; then
     fi
     AC_MSG_CHECKING([for extension for Fortran 90 programs])
     ac_f90ext="f90"
-    ac_f90compile='${FC-f90} -c $FCFLAGS conftest.$ac_f90ext 1>&AC_FD_CC'
+    ac_f90compile='${FC-f90} -c $FCFLAGS conftest.$ac_f90ext 1>&AS_MESSAGE_LOG_FD'
     cat > conftest.$ac_f90ext <<EOF
       program conftest
       end
diff --git a/confdb/aclocal_shl.m4 b/confdb/aclocal_shl.m4
index f974bfa..78bba26 100644
--- a/confdb/aclocal_shl.m4
+++ b/confdb/aclocal_shl.m4
@@ -57,11 +57,11 @@ dnl D*/
 AC_DEFUN([PAC_ARG_SHAREDLIBS],[
 
 AC_ARG_ENABLE(shared,
-	AC_HELP_STRING([--enable-shared], [Enable shared library builds]),,
+	AS_HELP_STRING([--enable-shared], [Enable shared library builds]),,
 	enable_shared=no)
 
 AC_ARG_ENABLE(rpath,
-	AC_HELP_STRING([--enable-rpath],
+	AS_HELP_STRING([--enable-rpath],
 		[Determine whether the rpath is set when programs are
 		compiled and linked when shared libraries are built.
 		The default is yes; use --disable-rpath to turn this
diff --git a/confdb/aclocal_shm.m4 b/confdb/aclocal_shm.m4
index d2cbaad..b4d2911 100644
--- a/confdb/aclocal_shm.m4
+++ b/confdb/aclocal_shm.m4
@@ -21,7 +21,7 @@ AC_DEFUN([PAC_ARG_SHARED_MEMORY],[
 
 # check how to allocate shared memory
 AC_ARG_WITH(shared-memory,
-    AC_HELP_STRING([--with-shared-memory[=auto|sysv|mmap]], [create shared memory using sysv or mmap (default is auto)]),,
+    AS_HELP_STRING([--with-shared-memory[=auto|sysv|mmap]], [create shared memory using sysv or mmap (default is auto)]),,
     with_shared_memory=auto)
 
 if test "$with_shared_memory" = auto -o "$with_shared_memory" = mmap; then
diff --git a/src/mpi/romio/autogen.sh b/src/mpi/romio/autogen.sh
index 9d01f53..df86e50 100755
--- a/src/mpi/romio/autogen.sh
+++ b/src/mpi/romio/autogen.sh
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-${AUTORECONF:-autoreconf} ${autoreconf_args:-"-vif"} -I confdb
+${AUTORECONF:-autoreconf} ${autoreconf_args:-"-vif"} -I confdb --warnings=all
 
 echo "=== running autoreconf in 'mpl' ==="
 (cd mpl && ${AUTORECONF:-autoreconf} ${autoreconf_args:-"-vif"}) || exit 1
diff --git a/src/mpi/romio/configure.ac b/src/mpi/romio/configure.ac
index 67732c4..c197306 100644
--- a/src/mpi/romio/configure.ac
+++ b/src/mpi/romio/configure.ac
@@ -441,7 +441,7 @@ fi
 
 PAC_GET_SPECIAL_SYSTEM_INFO
 
-AC_HAVE_FUNCS(memalign)
+AC_CHECK_FUNC(memalign)
 
 #
 # Question: Should ROMIO under MPICH ignore the Fortran tests, since 
@@ -700,22 +700,38 @@ AM_CONDITIONAL([BUILD_F77_TESTS],[test "x$NOF77" != "x1"])
 #
 # Check whether the MPI Offset type is compatible with struct flock
 AC_MSG_CHECKING([whether struct flock compatible with MPI_Offset])
-AC_TRY_COMPILE([#include <fcntl.h>],
-[struct flock l;
- $MPI_OFFSET_TYPE a=1;
- l.l_start = a; 
- l.l_len   = a;
-],pac_cv_struct_flock_and_mpi_offset=yes,pac_cv_struct_flock_and_mpi_offset=no)
+AC_COMPILE_IFELSE([
+    AC_LANG_SOURCE([
+#include <fcntl.h>
+
+int main(int argc, char **argv) {
+    struct flock l;
+    $MPI_OFFSET_TYPE a=1;
+
+    l.l_start = a;
+    l.l_len   = a;
+    return 0;
+}
+])
+],[pac_cv_struct_flock_and_mpi_offset=yes],[pac_cv_struct_flock_and_mpi_offset=no])
 AC_MSG_RESULT($pac_cv_struct_flock_and_mpi_offset)
 # FIXME: We should look for struct flock64 and the F_SETLK64/F_GETLK64
 # ADIOI_Set_lock could use these instead.
 if test "$pac_cv_struct_flock_and_mpi_offset" = no ; then
     AC_MSG_CHECKING([whether struct flock compatible with int])
-    AC_TRY_COMPILE([#include <fcntl.h>],
-[struct flock l;
- int a=1;
- l.l_start = a; 
- l.l_len   = a;
+    AC_COMPILE_IFELSE([
+        AC_LANG_SOURCE([
+#include <fcntl.h>
+
+int main(int argc, char **argv) {
+    struct flock l;
+    int a=1;
+
+    l.l_start = a;
+    l.l_len   = a;
+    return 0;
+}
+])
 ],pac_cv_struct_flock_and_int=yes,pac_cv_struct_flock_and_int=no)
     AC_MSG_RESULT($pac_cv_struct_flock_and_int)
     if test "$pac_cv_struct_flock_and_int" = yes ; then
@@ -1150,8 +1166,11 @@ AS_CASE([$host_os],
 #
 AC_CHECK_HEADERS(sys/vfs.h sys/param.h sys/mount.h sys/statvfs.h)
 AC_CHECK_FUNCS([statfs])
-AC_MSG_CHECKING([whether struct statfs properly defined])
-AC_TRY_COMPILE([
+
+AC_CHECK_TYPE([struct statfs],
+              [pac_cv_have_statfs=yes],
+              [pac_cv_have_statfs=no],
+              [AC_INCLUDES_DEFAULT
 #ifdef HAVE_SYS_VFS_H
 #include <sys/vfs.h>
 #endif
@@ -1164,8 +1183,6 @@ AC_TRY_COMPILE([
 #ifdef HAVE_SYS_MOUNT_H
 #include <sys/mount.h>
 #endif
-    ],[
-    struct statfs f;
     ],
     pac_cv_have_statfs=yes,pac_cv_have_statfs=no
 )
@@ -1177,8 +1194,10 @@ AC_MSG_RESULT($pac_cv_have_statfs)
 if test "$pac_cv_have_statfs" = yes ; then
     AC_DEFINE(HAVE_STRUCT_STATFS,1,[Define if struct statfs can be compiled])
 fi
-AC_MSG_CHECKING([for f_type member of statfs structure])
-AC_TRY_COMPILE([
+AC_CHECK_MEMBER([struct statfs.f_type],
+                [AC_DEFINE(ROMIO_HAVE_STRUCT_STATFS_WITH_F_TYPE, 1,[Define if statfs has f_type])],
+                [],
+                [AC_INCLUDES_DEFAULT
 #ifdef HAVE_SYS_VFS_H
 #include <sys/vfs.h>
 #endif
@@ -1194,22 +1213,13 @@ AC_TRY_COMPILE([
 #ifdef HAVE_STRING_H
 #include <string.h>
 #endif
-    ],[
-    struct statfs f;
-    memset(&f, 0, sizeof(f));
-    f.f_type = 0;
-    ],
-    pac_cv_have_statfs_f_type=yes,
-    pac_cv_have_statfs_f_type=no
+    ]
 )
-AC_MSG_RESULT($pac_cv_have_statfs_f_type)
-if test $pac_cv_have_statfs_f_type = yes ; then
-    AC_DEFINE(ROMIO_HAVE_STRUCT_STATFS_WITH_F_TYPE, 1,[Define if statfs has f_type])
-fi
 
-
-AC_MSG_CHECKING([for f_fstypename member of statfs structure])
-AC_TRY_COMPILE([
+AC_CHECK_MEMBER([struct statfs.f_fstypename],
+                [AC_DEFINE(ROMIO_HAVE_STRUCT_STATFS_WITH_F_FSTYPENAME, 1,[Define if statfs has f_fstypename])],
+                [],
+                [AC_INCLUDES_DEFAULT
 #ifdef HAVE_SYS_VFS_H
 #include <sys/vfs.h>
 #endif
@@ -1224,19 +1234,7 @@ AC_TRY_COMPILE([
 #endif
 #ifdef HAVE_STRING_H
 #include <string.h>
-#endif
-    ],[
-    struct statfs f;
-    memset(&f, 0, sizeof(f));
-    strncmp("nfs", f.f_fstypename, 3);
-    ],
-    pac_cv_have_statfs_f_fstypename=yes,
-    pac_cv_have_statfs_f_fstypename=no
-)
-AC_MSG_RESULT($pac_cv_have_statfs_f_fstypename)
-if test $pac_cv_have_statfs_f_fstypename = yes ; then
-    AC_DEFINE(ROMIO_HAVE_STRUCT_STATFS_WITH_F_FSTYPENAME, 1,[Define if statfs has f_fstypename])
-fi
+#endif])
 
 #
 # Check for stat and st_fstype field (NEC SX4)
@@ -1244,8 +1242,10 @@ fi
 AC_CHECK_HEADERS(sys/stat.h sys/types.h unistd.h)
 AC_CHECK_FUNCS(stat,
     AC_DEFINE(HAVE_STAT, 1, Define if stat function is present)
-    AC_MSG_CHECKING([for st_fstype member of stat structure])
-    AC_TRY_COMPILE([
+    AC_CHECK_MEMBER([struct stat.st_fstype],
+                    [AC_DEFINE(ROMIO_HAVE_STRUCT_STAT_WITH_ST_FSTYPE, 1, Define if struct stat has a st_fstype member)],
+                    [],
+                    [AC_INCLUDES_DEFAULT
 #ifdef HAVE_SYS_TYPES_H
 #include <sys/types.h>
 #endif
@@ -1255,14 +1255,7 @@ AC_CHECK_FUNCS(stat,
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>
 #endif
-	],[
-	struct stat s;
-
-	s.st_fstype = NULL;
 	],
-	AC_MSG_RESULT(yes)
-	AC_DEFINE(ROMIO_HAVE_STRUCT_STAT_WITH_ST_FSTYPE, 1, Define if struct stat has a st_fstype member),
-	AC_MSG_RESULT(no)
     )
 )
 
@@ -1272,8 +1265,10 @@ AC_CHECK_FUNCS(stat,
 AC_CHECK_HEADERS(sys/types.h sys/statvfs.h sys/vfs.h)
 AC_CHECK_FUNCS(statvfs,
     AC_DEFINE(HAVE_STATVFS, 1, Define if statvfs function is present)
-    AC_MSG_CHECKING([for f_basetype member of statvfs structure])
-    AC_TRY_COMPILE([
+    AC_CHECK_MEMBER([struct statvfs.f_basetype],
+                    [AC_DEFINE(ROMIO_HAVE_STRUCT_STATVFS_WITH_F_BASETYPE, 1, defined if struct statvfs has a f_basetype member)],
+                    [],
+                    [AC_INCLUDES_DEFAULT
 #ifdef HAVE_SYS_TYPES_H
 #include <sys/types.h>
 #endif
@@ -1283,14 +1278,7 @@ AC_CHECK_FUNCS(statvfs,
 #ifdef HAVE_SYS_STATVFS_H
 #include <sys/statvfs.h>
 #endif
-	], [[
-	struct statvfs foo;
-
-	foo.f_basetype[0] = 'a';
-	]],
-	AC_MSG_RESULT(yes)
-	AC_DEFINE(ROMIO_HAVE_STRUCT_STATVFS_WITH_F_BASETYPE, 1, defined if struct statvfs has a f_basetype member),
-	AC_MSG_RESULT(no)
+	]
     )
 )
 
@@ -1313,22 +1301,38 @@ AC_CHECK_TYPE([blksize_t],[],[AC_DEFINE_UNQUOTED([blksize_t],[__blksize_t],[Prov
 AC_CHECK_HEADERS(unistd.h)
 if test "$ac_cv_header_unistd_h" = "yes" ; then
     AC_CACHE_CHECK([for large file defines],ac_cv_has_large_file_defines,[
-    AC_TRY_COMPILE([
+    AC_COMPILE_IFELSE([
+        AC_LANG_SOURCE([
 #include <unistd.h>
-        ], [
+
 #ifndef _LFS64_LARGEFILE
 #error no largefile defines
 #endif
-        ],ac_cv_has_large_file_defines=yes,ac_cv_has_large_file_defines=no)])
+
+int main(int argc, char **argv) {
+    return 0;
+}
+])],
+                      [ac_cv_has_large_file_defines=yes],
+                      [ac_cv_has_large_file_defines=no])])
     if test "$ac_cv_has_large_file_defines" = "yes" ; then
         # See if we can use them
     AC_CACHE_CHECK([whether off64_t is an scalar type],ac_cv_off64_t_scalar,[
-    AC_TRY_COMPILE([
+    AC_COMPILE_IFELSE([
+        AC_LANG_SOURCE([
 #define _LARGEFILE64_SOURCE
 #define _FILE_OFFSET_BITS 64
 #include <sys/types.h>
 #include <unistd.h>
-],[off64_t a = 3;],ac_cv_off64_t_scalar=yes,ac_cv_off64_t_scalar=no)])
+
+off64_t a = 3;
+
+int main(int argc, char **argv) {
+    return 0;
+}
+])],
+                      [ac_cv_off64_t_scalar=yes],
+                      [ac_cv_off64_t_scalar=no])])
 
         if test "$ac_cv_off64_t_scalar" = "yes" ; then
             CFLAGS="${CFLAGS} -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64"
@@ -1635,16 +1639,21 @@ fi
 
 AC_CHECK_HEADERS(unistd.h,
         AC_MSG_CHECKING([for large file defines])
-        AC_TRY_COMPILE([
-                #include <unistd.h>
-                ], [
-                #ifndef _LFS64_LARGEFILE
-                #error no largefile defines
-                #endif
-                ],
-                CFLAGS="${CFLAGS} -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64"
-                AC_MSG_RESULT(yes),
-                AC_MSG_RESULT(none) )
+        AC_COMPILE_IFELSE([
+            AC_LANG_SOURCE([
+#include <unistd.h>
+
+#ifndef _LFS64_LARGEFILE
+#error no largefile defines
+#endif
+
+int main(int argc, char **argv) {
+    return 0;
+}
+                ])],
+                          [CFLAGS="${CFLAGS} -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64"
+                           AC_MSG_RESULT(yes)],
+                          [AC_MSG_RESULT(none)] )
         )
 
 echo "setting SYSDEP_INC to $SYSDEP_INC"
