diff --git a/src/mpid/ch3/channels/nemesis/nemesis/netmod/gm/gm_init.c b/src/mpid/ch3/channels/nemesis/nemesis/netmod/gm/gm_init.c
index eec8850..0f27c6c 100644
--- a/src/mpid/ch3/channels/nemesis/nemesis/netmod/gm/gm_init.c
+++ b/src/mpid/ch3/channels/nemesis/nemesis/netmod/gm/gm_init.c
@@ -268,6 +268,7 @@ MPID_nem_gm_vc_init(MPIDI_VC_t *vc)
     int mpi_errno = MPI_SUCCESS;
     int ret;
     char *business_card;
+    int pmi_errno;
     int val_max_sz;
     MPIU_CHKLMEM_DECL(1);
 
@@ -276,7 +277,7 @@ MPID_nem_gm_vc_init(MPIDI_VC_t *vc)
     MPIU_ERR_CHKANDJUMP1(pmi_errno, mpi_errno, MPI_ERR_OTHER, "**fail", "**fail %d", pmi_errno);
     MPIU_CHKLMEM_MALLOC(business_card, char *, val_max_sz, mpi_errno, "bc");
 
-    mpi_errno = vc->pg->getConnInfo(vc->pg_rank, business_card, key_max_sz, vc->pg);
+    mpi_errno = vc->pg->getConnInfo(vc->pg_rank, business_card, val_max_sz, vc->pg);
     if (mpi_errno) MPIU_ERR_POP(mpi_errno);
 
     VC_FIELD(vc, source_id) = my_pg_rank; /* FIXME: this is only valid for processes in COMM_WORLD */
diff --git a/src/mpid/ch3/channels/nemesis/nemesis/netmod/gm/gm_send.c b/src/mpid/ch3/channels/nemesis/nemesis/netmod/gm/gm_send.c
index d6ed0e9..9cd7fa1 100644
--- a/src/mpid/ch3/channels/nemesis/nemesis/netmod/gm/gm_send.c
+++ b/src/mpid/ch3/channels/nemesis/nemesis/netmod/gm/gm_send.c
@@ -13,8 +13,6 @@ send_callback (struct gm_port *p, void *context, gm_status_t status)
 {
     MPID_nem_cell_t *cell = (MPID_nem_cell_t *)context;
 
-    printf_d ("send_callback()\n");
-
     if (status != GM_SUCCESS)
     {
 	gm_perror ("Send error", status);
@@ -39,8 +37,6 @@ send_cell (int node_id, int port_id, MPID_nem_cell_ptr_t cell, int datalen)
     gm_send_with_callback (MPID_nem_module_gm_port, pkt, PACKET_SIZE, datalen + MPID_NEM_MPICH2_HEAD_LEN, GM_LOW_PRIORITY, node_id,
 			   port_id, send_callback, (void *)cell);
     DO_PAPI (PAPI_accum_var (PAPI_EventSet, PAPI_vvalues4));
-    printf_d ("  Sent packet to node = %d, port = %d\n", node_id, port_id);
-    printf_d ("    datalen %d\n", datalen);
     return MPI_SUCCESS;
 }
 
