From ecd35cffcf5b3659bb320c1a9475318337dbab2f Mon Sep 17 00:00:00 2001
From: Jeff Hammond <jeff.science@gmail.com>
Date: Fri, 21 Feb 2014 13:31:30 -0600
Subject: [PATCH 2/2] change default for abort msg suppression to false

---
 src/mpi/init/abort.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/mpi/init/abort.c b/src/mpi/init/abort.c
index 24de3544..f0b4cdc8 100644
--- a/src/mpi/init/abort.c
+++ b/src/mpi/init/abort.c
@@ -17,7 +17,7 @@ cvars:
     - name        : MPIR_CVAR_SUPPRESS_ABORT_MESSAGE
       category    : ERROR_HANDLING
       type        : boolean
-      default     : true
+      default     : false
       class       : device
       verbosity   : MPI_T_VERBOSITY_USER_BASIC
       scope       : MPI_T_SCOPE_ALL_EQ
@@ -129,8 +129,8 @@ int MPI_Abort(MPI_Comm comm, int errorcode)
     {
 	MPIU_Snprintf(comm_name, MPI_MAX_OBJECT_NAME, "comm=0x%X", comm);
     }
-    /* FIXME: This is not internationalized */
     if (!MPIR_CVAR_SUPPRESS_ABORT_MESSAGE)
+        /* FIXME: This is not internationalized */
         MPIU_Snprintf(abort_str, 100, "application called MPI_Abort(%s, %d) - process %d", comm_name, errorcode, comm_ptr->rank);
     mpi_errno = MPID_Abort( comm_ptr, mpi_errno, errorcode, abort_str );
     /* --BEGIN ERROR HANDLING-- */
-- 
1.8.5.3

