Index: maint/simplemake.in
===================================================================
--- maint/simplemake.in	(revision 4165)
+++ maint/simplemake.in	(working copy)
@@ -418,6 +418,7 @@
 		     'HTML'    => '$(INSTALL_DATA)',
 		     'DOC'     => '$(INSTALL_DATA)',
 		     'ETC'     => '$(INSTALL_DATA)',
+		     'DATADIR' => '$(INSTALL_DATA)',
 		     'PKGCONFIG' => '$(INSTALL_DATA)',
 		    );
 %uninstall_methods = ( 'SHLIB'   => '$(CREATESHLIB) --mode=uninstall',
