Index: src/mpid/common/locks/mpidu_atomic_primitives.h
===================================================================
--- src/mpid/common/locks/mpidu_atomic_primitives.h	(revision 3981)
+++ src/mpid/common/locks/mpidu_atomic_primitives.h	(working copy)
@@ -455,7 +455,7 @@
 #elif defined(HAVE_GCC_AND_IA64_ASM)
     /* is pointer swizzling necessary here? */
     __asm__ __volatile__ ("xchg8 %0=[%2],%3"
-                          : "=r" (val), "=m" (*val)
+                          : "=r" (val), "=m" (*ptr)
                           : "r" (ptr), "0" (val));
     return val;
 #else /* "lock-op-unlock" fallback */
@@ -501,7 +501,7 @@
 #elif defined(HAVE_GCC_AND_IA64_ASM)
     /* is pointer swizzling necessary here? */
     __asm__ __volatile__ ("xchg8 %0=[%2],%3"
-                          : "=r" (val), "=m" (*val)
+                          : "=r" (val), "=m" (*ptr)
                           : "r" (ptr), "0" (val));
     return val;
 #else /* "lock-op-unlock" fallback */
@@ -535,7 +535,7 @@
 #elif defined(HAVE_GCC_AND_IA64_ASM)
     /* is pointer swizzling necessary here? */
     __asm__ __volatile__ ("xchg8 %0=[%2],%3"
-                          : "=r" (val), "=m" (*val)
+                          : "=r" (val), "=m" (*ptr)
                           : "r" (ptr), "0" (val));
     return val;
 #else /* "lock-op-unlock" fallback */
